File: two_sigma/financial_modeling_challenge/cpp/src/data_generator.cpp
#include "data_generator.h"
#include <fstream>
#include <iostream>
#include <random>
#include <filesystem>

namespace two_sigma {

DataGenerator::DataGenerator(int n_samples, int n_features, int n_instruments)
    : n_samples_(n_samples), n_features_(n_features), n_instruments_(n_instruments) {}

void DataGenerator::generate(const std::string& output_path) {
    std::filesystem::create_directories(std::filesystem::path(output_path).parent_path());
    std::ofstream out(output_path);
    if (!out) return;

    // Header
    out << "id,timestamp,y";
    for (int i = 0; i < n_features_; ++i) {
        out << ",technical_" << i;
    }
    out << "\n";

    std::mt19937 gen(42);
    std::normal_distribution<double> dist_y(0.0, 0.02);
    std::normal_distribution<double> dist_feat(0.0, 1.0);

    int n_timestamps = n_samples_ / n_instruments_;
    for (int t = 0; t < n_timestamps; ++t) {
        for (int i = 0; i < n_instruments_; ++i) {
            out << i << "," << t << "," << dist_y(gen);
            for (int f = 0; f < n_features_; ++f) {
                out << "," << dist_feat(gen);
            }
            out << "\n";
        }
    }
    std::cout << "Generated synthetic data at " << output_path << std::endl;
}

} // namespace two_sigma

File: two_sigma/financial_modeling_challenge/rust/src/data_generator.rs
use std::fs::File;
use std::io::Write;
use rand::prelude::*;
use rand_distr::{Normal, Distribution};

pub struct DataGenerator {
    pub n_samples: usize,
    pub n_features: usize,
    pub n_instruments: usize,
}

impl DataGenerator {
    pub fn new(n_samples: usize, n_features: usize, n_instruments: usize) -> Self {
        Self { n_samples, n_features, n_instruments }
    }

    pub fn generate(&self, output_path: &str) -> std::io::Result<()> {
        let mut file = File::create(output_path)?;
        
        // Header
        write!(file, "id,timestamp,y")?;
        for i in 0..self.n_features {
            write!(file, ",technical_{}", i)?;
        }
        writeln!(file)?;

        let mut rng = StdRng::seed_from_u64(42);
        let dist_y = Normal::new(0.0, 0.02).unwrap();
        let dist_feat = Normal::new(0.0, 1.0).unwrap();

        let n_timestamps = self.n_samples / self.n_instruments;
        for t in 0..n_timestamps {
            for i in 0..self.n_instruments {
                write!(file, "{},{},{}", i, t, dist_y.sample(&mut rng))?;
                for _ in 0..self.n_features {
                    write!(file, ",{}", dist_feat.sample(&mut rng))?;
                }
                writeln!(file)?;
            }
        }
        println!("Generated synthetic data at {}", output_path);
        Ok(())
    }
}
